pyanaconda/storage_utils.py

```diff
@@ -845,3 +847,77 @@ def device_matches(spec, devicetree=None, disks_only=False):
+
+    Updates "nvdimm use" commands.  Doesn't add use command for devices which
+    are reconfigured with "nvdimm reconfigure" because reconfigure in kickstart
+    implies use.
+    """
+    if namespaces is None:
+        namespaces_to_add = set()
+    else:
+        namespaces_to_add = set(namespaces)
+
+    new_actionList = []
+
+    # Keep all commands except for use, track reconfigured namespaces
+    reconfigured_namespaces = set()
+    for nvdimm_command in data.nvdimm.actionList:
+        if nvdimm_command.action == NVDIMM_ACTION_USE:
+            continue
+        if nvdimm_command.action == NVDIMM_ACTION_RECONFIGURE:
+            reconfigured_namespaces.add(nvdimm_command.namespace)
+        new_actionList.append(nvdimm_command)
+
+    # Reconfigured namespaces are used implicitly, don't add them as use
+    namespaces_to_add.difference_update(reconfigured_namespaces)
+
+    # Add use commands
+    for ns in sorted(namespaces_to_add):
+        added_nvdimm_command = _nvdimm_create_ksdata(action=NVDIMM_ACTION_USE,
+                                                     namespace=ns)
+        new_actionList.append(added_nvdimm_command)
+
+    data.nvdimm.actionList = new_actionList
+
+
+def nvdimm_update_ksdata_after_reconfiguration(data, namespace, mode=NVDIMM_MODE_SECTOR,
+                                               sectorsize=None):
+    """Update ks data after reconfiguration of NVDIMM device.
+
+    Updates "nvdimm reconfigure" commands.
+    """
+    for nvdimm_command in data.nvdimm.actionList:
+        # If use or reconfigure already exists, modify it
+        if namespace and nvdimm_command.namespace == namespace and \
+                nvdimm_command.action in [NVDIMM_ACTION_RECONFIGURE,
+                                          NVDIMM_ACTION_USE]:
+            nvdimm_command.mode = mode
+            nvdimm_command.sectorsize = sectorsize
+            nvdimm_command.action = NVDIMM_ACTION_RECONFIGURE
+            break
+    else:
+        # If neither use nor reconfigure already exists, add it
+        added_nvdimm_command = _nvdimm_create_ksdata(action=NVDIMM_ACTION_RECONFIGURE,
+                                                     namespace=namespace,
+                                                     sectorsize=sectorsize,
+                                                     mode=mode)
+        data.nvdimm.actionList.append(added_nvdimm_command)
+
+
+def _nvdimm_create_ksdata(action=None, namespace=None, mode=None, sectorsize=None):
+    from pyanaconda.kickstart import AnacondaKSHandler
+    handler = AnacondaKSHandler()
+    # pylint: disable=E1101
+    nvdimm_data = handler.NvdimmData()
+    if action is not None:
+        nvdimm_data.action = action
+    if namespace is not None:
+        nvdimm_data.namespace = namespace
+    if mode is not None:
+        nvdimm_data.mode = mode
+    if sectorsize is not None:
+        nvdimm_data.sectorsize = sectorsize
+    return nvdimm_data
```

Comment:
I'm not a fan of adding the code here, we have more places like this storage_utils and it grows and grows. After some time no-one knows what is inside. I would recommend to tight these together to separate them from the rest of the non-nvdimm code here.
You can separate these to their own file or at least couple them to a class.

Reply:
I am not sure if it is worth it before modularization, this will all probably go to storage module (nvdimm submodule, its kickstart.py ?) anyway. Actually the code is tight more about UI action / ks update than about storage. For iscsi similar code is in the GUI spoke. I think the only reason I moved the code for nvdimm elsewhere was that I wanted to write unit tests for the functions.
Do you have any suggestion? To me figuring this out now seems like a waste of time.

Reply:
True that the modularization can change this from the bottom. OK in that case, you can ignore this.